# 字典


In [ ]:
#使用花括号创建字典
data = {'name':'Alice','age':25,'city':'New York'}
data

{'name': 'Alice', 'age': 25, 'city': 'New York'}

In [ ]:
print(data['age'])

25


# 标签编码
之前介绍了离散数据 如果是不存在顺序，则采用独热编码，函数为pd.get_dummies()

现在介绍对于存在顺序和大小关系的离散特征，做好标签编码,借助dataframe的map函数即可实现

In [18]:
import pandas as pd 
data = pd.read_csv(r'data.csv')
data.head()

Id Home Ownership  Annual Income Years in current job  Tax Liens  \
0   0       Own Home       482087.0                  NaN        0.0   
1   1       Own Home      1025487.0            10+ years        0.0   
2   2  Home Mortgage       751412.0              8 years        0.0   
3   3       Own Home       805068.0              6 years        0.0   
4   4           Rent       776264.0              8 years        0.0   

   Number of Open Accounts  Years of Credit History  Maximum Open Credit  \
0                     11.0                     26.3             685960.0   
1                     15.0                     15.3            1181730.0   
2                     11.0                     35.0            1182434.0   
3                      8.0                     22.5             147400.0   
4                     13.0                     13.6             385836.0   

   Number of Credit Problems  Months since last delinquent  Bankruptcies  \
0                        1.0                           NaN           1.0   
1                        0.0                           NaN           0.0   
2                        0.0                           NaN           0.0   
3                        1.0                           NaN           1.0   
4                        1.0                           NaN           0.0   

              Purpose        Term  Current Loan Amount  \
0  debt consolidation  Short Term           99999999.0   
1  debt consolidation   Long Term             264968.0   
2  debt consolidation  Short Term           99999999.0   
3  debt consolidation  Short Term             121396.0   
4  debt consolidation  Short Term             125840.0   

   Current Credit Balance  Monthly Debt  Credit Score  Credit Default  
0                 47386.0        7914.0         749.0               0  
1                394972.0       18373.0         737.0               1  
2                308389.0       13651.0         742.0               0  
3                 95855.0       11338.0         694.0               0  
4                 93309.0        7180.0         719.0               0

In [10]:
data['Home Ownership'].value_counts() # 查看一下Home Ownership的取值

Home Ownership
Home Mortgage    3637
Rent             3204
Own Home          647
Have Mortgage      12
Name: count, dtype: int64

* 住房抵押贷款：3637 这个是有房贷，有房子
* 租房：3204 没房子
* 拥有自有住房：647 这个没贷款，有房子
* 有贷款：12 这个是有其他贷款，有房子，没房贷
* 名称：房屋所有权，数据类型：int64
上面这些翻译也不一定对，这是我个人对数据的理解。

按照贷款严重程度（抗风险能力），依次是：自有住房 ＜ 租房 ＜ 有其他贷款 ＜ 住房抵押贷款

所以按照这个逻辑来进行编码

In [19]:
# 可以看出'Home Ownership'是一个离散变量
# 定义映射字典
mapping_dict = {
    "Rent":0,
    "Own Home":1,
    "Have Mortgage":2,
    "Home Mortgage":3
}
data['Home Ownership'].head()

0         Own Home
1         Own Home
2    Home Mortgage
3         Own Home
4             Rent
Name: Home Ownership, dtype: object

In [20]:
data["Home Ownership"] = data["Home Ownership"].map(mapping_dict)
data["Home Ownership"].head()

0    1
1    1
2    3
3    1
4    0
Name: Home Ownership, dtype: int64

In [21]:
data["Term"].value_counts()

Term
Short Term    5556
Long Term     1944
Name: count, dtype: int64

对于字符串类型 我们也要映射成整数类型，这里不要理解为标签编码或者独热编码

二分类的问题不需要独热编码，比如性别这个特征，男女不需要变成2个特征，性别男 性别女 。因为他们二者自由度为1，如果是2个特征的话，性别男=1，那么性别女必定等于0.这样特征高度相关，没有价值。

此时这个特征的含义不是性别，而是：是否为男性，1是男性，0是非男。

三分类以上才涉及独热编码

In [22]:
# 定义映射字典
mapping_dict = {
    "Short Term" : 0,
    "Long Term" : 1
}
data["Term"] = data["Term"].map(mapping_dict)
data["Term"].head()

0    0
1    1
2    0
3    0
4    0
Name: Term, dtype: int64

有个骚操作，实际上借助一个映射函数也可以实现上面2次编码

#### 字典的键值对可以嵌套字典

In [23]:
import pandas as pd

# 重新读取数据
data = pd.read_csv("data.csv")
# 嵌套映射字典
mapping = {
    "Term": {
        "Short Term": 1,
        "Long Term": 0
    },
    "Home Ownership": {
        "Rent": 0,
        "Own Home": 1,
        "Have Mortgage  ": 2,
        "Home Mortgage": 3
    }
}

In [24]:
mapping["Term"] # 访问嵌套字典中的值，此时他又是一个字典

{'Short Term': 1, 'Long Term': 0}

In [25]:
# 对 Home Ownership 列进行映射
data["Home Ownership"] = data["Home Ownership"].map(mapping["Home Ownership"])

# 对 Term 列进行映射
data["Term"] = data["Term"].map(mapping["Term"])

data.head()

Id  Home Ownership  Annual Income Years in current job  Tax Liens  \
0   0             1.0       482087.0                  NaN        0.0   
1   1             1.0      1025487.0            10+ years        0.0   
2   2             3.0       751412.0              8 years        0.0   
3   3             1.0       805068.0              6 years        0.0   
4   4             0.0       776264.0              8 years        0.0   

   Number of Open Accounts  Years of Credit History  Maximum Open Credit  \
0                     11.0                     26.3             685960.0   
1                     15.0                     15.3            1181730.0   
2                     11.0                     35.0            1182434.0   
3                      8.0                     22.5             147400.0   
4                     13.0                     13.6             385836.0   

   Number of Credit Problems  Months since last delinquent  Bankruptcies  \
0                        1.0                           NaN           1.0   
1                        0.0                           NaN           0.0   
2                        0.0                           NaN           0.0   
3                        1.0                           NaN           1.0   
4                        1.0                           NaN           0.0   

              Purpose  Term  Current Loan Amount  Current Credit Balance  \
0  debt consolidation     1           99999999.0                 47386.0   
1  debt consolidation     0             264968.0                394972.0   
2  debt consolidation     1           99999999.0                308389.0   
3  debt consolidation     1             121396.0                 95855.0   
4  debt consolidation     1             125840.0                 93309.0   

   Monthly Debt  Credit Score  Credit Default  
0        7914.0         749.0               0  
1       18373.0         737.0               1  
2       13651.0         742.0               0  
3       11338.0         694.0               0  
4        7180.0         719.0               0

以上是对离散变量的处理方法。
下面是对连续变量的处理方法：

# 连续变量的处理
归一化和标准化可以通过手写函数实现，也可以使用sklearn中的归一化和标准化函数。

归一化通常是指将数据缩放到特定的范围内，通常是将数据缩放到0到1之间。归一化的目的是为了消除不同特征之间的量纲影响，使得不同特征之间的权重相等。归一化的方法有很多种，常见的有最小-最大归一化、Z-score归一化、小数定标归一化等。

![alt text](<屏幕截图 2025-04-27 233534-1.png>)

![alt text](<屏幕截图 2025-04-27 233647.png>)

#### 方法一：手动实现归一化

In [ ]:
# 对"Annual Income (k$)"列进行归一化处理
#下面是一个函数，用于对指定列进行归一化处理：
def normalize_column(column):
    min_val = column.min()
    max_val = column.max()
    normalized_column = (column - min_val) / (max_val - min_val) # 上面图片里的归一化公式
    return normalized_column
# 调用函数进行归一化处理
data['Annual Income'] = normalize_column(data['Annual Income']) # 对表中的"Annual Income"列进行归一化处理
data['Annual Income'].head()

0    0.031798
1    0.086221
2    0.058771
3    0.064145
4    0.061260
Name: Annual Income, dtype: float64

#### 方法二：调用sklearn库进行归一化

In [28]:
#调用sklearn库进行归一化
from sklearn.preprocessing import MinMaxScaler,StandardScaler
data = pd.read_csv(r'data.csv')
# 归一化处理
scaler = MinMaxScaler() # 实例化
data['Annual Income'] = scaler.fit_transform(data[['Annual Income']]) # 归一化处理
data['Annual Income'].head()


0    0.031798
1    0.086221
2    0.058771
3    0.064145
4    0.061260
Name: Annual Income, dtype: float64

#### 标准化
标准化和归一化的区别：
1. 归一化是将数据缩放到指定的范围内，通常是[0,1]或[-1,1]。
2. 标准化是将数据转换为均值为0，方差为1的分布，可以理解为[-1,1]。

In [29]:
# 标准化处理
data = pd.read_csv(r'data.csv')
scaler = StandardScaler()
data['Annual Income'] = scaler.fit_transform(data[['Annual Income']])
data['Annual Income'].head()

0   -1.046183
1   -0.403310
2   -0.727556
3   -0.664078
4   -0.698155
Name: Annual Income, dtype: float64